File: src/lib/messages/commandcode.ts
import fs from 'node:fs/promises'
import type { ChatMessage, ChatMessageRole } from '../types'
import { extractTextFromParts } from './extract-text'

interface ContentPart {
  type?: string
  text?: string
  name?: string
  toolName?: string
  input?: unknown
}

function nonemptyString(value: unknown): string | undefined {
  if (typeof value !== 'string') return undefined
  const trimmed = value.trim()
  return trimmed ? trimmed : undefined
}

function textFromContent(content: unknown): string {
  if (typeof content === 'string') return content.trim()
  if (Array.isArray(content)) {
    return extractTextFromParts(content as Array<{ type?: string; text?: string }>)
  }
  return ''
}

function formatToolUse(part: ContentPart): string {
  const name = part.name?.trim() || part.toolName?.trim() || 'tool'
  const input =
    part.input && typeof part.input === 'object'
      ? JSON.stringify(part.input, null, 2)
      : ''
  return input ? `${name}\n${input}` : name
}

function parseEntry(
  row: Record<string, unknown>,
): { role: string; content: unknown; timestamp?: string } | null {
  const timestamp = nonemptyString(row.timestamp)
  if (row.type === 'message' && row.message && typeof row.message === 'object') {
    const message = row.message as Record<string, unknown>
    const role = nonemptyString(message.role)
    if (!role) return null
    return { role, content: message.content, timestamp }
  }
  const role = nonemptyString(row.role)
  if (!role) return null
  return { role, content: row.content, timestamp }
}

function assistantParts(
  content: unknown,
): Array<{ role: ChatMessageRole; content: string }> {
  if (typeof content === 'string') {
    const trimmed = content.trim()
    return trimmed ? [{ role: 'assistant', content: trimmed }] : []
  }
  if (!Array.isArray(content)) return []
  const messages: Array<{ role: ChatMessageRole; content: string }> = []
  for (const part of content as ContentPart[]) {
    if (part.type === 'text' && part.text?.trim()) {
      messages.push({ role: 'assistant', content: part.text.trim() })
      continue
    }
    if (part.type === 'tool_use' || part.type === 'tool-call') {
      messages.push({ role: 'tool', content: formatToolUse(part) })
    }
  }
  return messages
}

export async function fetchCommandCodeMessages(
  sessionPath: string,
): Promise<ChatMessage[]> {
  try {
    const raw = await fs.readFile(sessionPath, 'utf-8')
    const messages: ChatMessage[] = []
    let index = 0
    for (const line of raw.split('\n')) {
      if (!line.trim()) continue
      try {
        const row = JSON.parse(line) as Record<string, unknown>
        const entry = parseEntry(row)
        if (!entry) continue
        appendEntry(messages, () => index++, entry)
      } catch {
        // skip malformed line
      }
    }
    return messages
  } catch {
    return []
  }
}

function appendEntry(
  messages: ChatMessage[],
  nextIndex: () => number,
  entry: { role: string; content: unknown; timestamp?: string },
) {
  if (entry.role === 'user') {
    const text = textFromContent(entry.content)
    if (!text) return
    messages.push({
      id: `commandcode-msg-${nextIndex()}`,
      role: 'user',
      content: text,
      timestamp: entry.timestamp,
    })
    return
  }
  if (entry.role !== 'assistant') return
  for (const part of assistantParts(entry.content)) {
    messages.push({
      id: `commandcode-msg-${nextIndex()}`,
      role: part.role,
      content: part.content,
      timestamp: entry.timestamp,
    })
  }
}

File: crates/ai-chats-core/src/messages/commandcode.rs
use crate::text::extract_text_from_parts;
use crate::types::{ChatMessage, ChatMessageRole};
use serde_json::Value;
use std::fs;
use std::path::Path;

fn nonempty_string(value: Option<&Value>) -> Option<String> {
    value
        .and_then(|v| v.as_str())
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
}

fn text_from_content(content: &Value) -> String {
    if let Some(s) = content.as_str() {
        return s.trim().to_string();
    }
    if let Some(arr) = content.as_array() {
        return extract_text_from_parts(arr);
    }
    String::new()
}

fn format_tool_use(part: &Value) -> String {
    let name = part
        .get("name")
        .or_else(|| part.get("toolName"))
        .and_then(|v| v.as_str())
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .unwrap_or("tool");
    let input = part.get("input").filter(|v| v.is_object());
    match input {
        Some(obj) => match serde_json::to_string_pretty(obj) {
            Ok(pretty) => format!("{name}\n{pretty}"),
            Err(_) => name.to_string(),
        },
        None => name.to_string(),
    }
}

fn parse_entry(row: &Value) -> Option<(String, Value, Option<String>)> {
    let timestamp = nonempty_string(row.get("timestamp"));
    if row.get("type").and_then(|v| v.as_str()) == Some("message") {
        let message = row.get("message")?;
        return Some((
            nonempty_string(message.get("role"))?,
            message.get("content").cloned().unwrap_or(Value::Null),
            timestamp,
        ));
    }
    Some((
        nonempty_string(row.get("role"))?,
        row.get("content").cloned().unwrap_or(Value::Null),
        timestamp,
    ))
}

fn assistant_parts(content: &Value) -> Vec<(ChatMessageRole, String)> {
    if let Some(s) = content.as_str() {
        let trimmed = s.trim();
        if trimmed.is_empty() {
            return Vec::new();
        }
        return vec![(ChatMessageRole::Assistant, trimmed.to_string())];
    }
    let Some(parts) = content.as_array() else {
        return Vec::new();
    };
    let mut messages = Vec::new();
    for part in parts {
        push_assistant_part(part, &mut messages);
    }
    messages
}

fn push_assistant_part(part: &Value, messages: &mut Vec<(ChatMessageRole, String)>) {
    let part_type = part.get("type").and_then(|v| v.as_str()).unwrap_or("");
    if part_type == "text" {
        if let Some(text) = part.get("text").and_then(|v| v.as_str()) {
            let trimmed = text.trim();
            if !trimmed.is_empty() {
                messages.push((ChatMessageRole::Assistant, trimmed.to_string()));
            }
        }
        return;
    }
    if part_type == "tool_use" || part_type == "tool-call" {
        messages.push((ChatMessageRole::Tool, format_tool_use(part)));
    }
}

fn push_message(
    messages: &mut Vec<ChatMessage>,
    index: &mut u32,
    role: ChatMessageRole,
    content: String,
    timestamp: Option<String>,
) {
    messages.push(ChatMessage {
        id: format!("commandcode-msg-{index}"),
        role,
        content,
        timestamp,
    });
    *index += 1;
}

/// Parse Command Code session `.jsonl` (v3 header+message or legacy v2 rows).
pub fn fetch_commandcode_messages(session_path: &Path) -> Vec<ChatMessage> {
    let content = match fs::read_to_string(session_path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };
    let mut messages = Vec::new();
    let mut index: u32 = 0;
    for line in content.lines() {
        if line.trim().is_empty() {
            continue;
        }
        let Ok(row) = serde_json::from_str::<Value>(line) else {
            continue;
        };
        let Some((role, content, timestamp)) = parse_entry(&row) else {
            continue;
        };
        append_entry(&mut messages, &mut index, &role, &content, timestamp);
    }
    messages
}

fn append_entry(
    messages: &mut Vec<ChatMessage>,
    index: &mut u32,
    role: &str,
    content: &Value,
    timestamp: Option<String>,
) {
    if role == "user" {
        let text = text_from_content(content);
        if text.is_empty() {
            return;
        }
        push_message(messages, index, ChatMessageRole::User, text, timestamp);
        return;
    }
    if role != "assistant" {
        return;
    }
    for (part_role, part_content) in assistant_parts(content) {
        push_message(messages, index, part_role, part_content, timestamp.clone());
    }
}
